In [ ]:
!pip install -q xlrd
!git clone https://github.com/biacraciun/Understanding-Context-Usage-in-MT

In [ ]:
!pip install -q -r Understanding-Context-Usage-in-MT/requirements.txt

In [ ]:
#Restart the session before executing this block of code so that the dependencies are loaded properly

import sys

# Add the path to the folder containing the modules
sys.path.append('Understanding-Context-Usage-in-MT/modules')

import setup_data_pecore
import use_pecore
import run_pecore

In [ ]:
from run_pecore import run_pecore

In [ ]:
run_pecore(langugage_code='ron_Latn', notebook=True)

config.json:   0%|          | 0.00/846 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/2.46G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/189 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/564 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/4.85M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.3M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/3.55k [00:00<?, ?B/s]

Context with contextual cues (std λ=2.00, top 5) followed by output sentence with context-sensitive target spans 
(all)
(CTI = "kl_divergence", CCI = "saliency" w/ "probability" target)

Input context:  Ecuador requires that Cuban citizens receive an invitation letter prior to entering Ecuador through
international airports or frontier admission points. This letter must be legalized by the Ecuadorian Ministry of 
Foreign Affairs, and comply with certain requirements. These requirements are designed to provide an organized 
migratory flow between both countries. Cuban citizens who are US green card holders should visit an Ecuadorian 
Consulate to obtain an exemption to this requirement.
Input current: Your passport must be valid for at least 6 months beyond your travel dates. A round/onward trip 
ticket is needed to prove the length of your stay.
Output context: Ecuadorul cere ca cetățenii Cubei să primească o scrisoare de invitație înainte de a intra în 
Ecuador folosind aeroporturile internaționale sau punctele de trecere a frontierei. Scrisoarea aceasta trebuie să 
fie legalizată de către Ministerul Afacerilor Externe al Ecuadorului și trebuie să respecte anumite cerințe. 
Cerințele acestea sunt concepute pentru a asigura o deplasare bine organizată între cele două țări. Cetățenii 
cubanezi care dețin cărți verzi pentru SUA trebuie să viziteze un consulat ecuadorian pentru a obține o scutire de 
la această cerință.
Output current: Pașaportul dumneavoastră trebuie să fie valabil pentru cel puțin 6 luni după datele de călătorie. 
Un bilet de călătorie în direcție / în continuare este necesar pentru a dovedi durata șederii dumneavoastră.

#1.
Generated output (CTI > 0.165):  Pa(1.870)șaportul dumneavoastră trebuie să fie valabil pentru cel puțin 6 luni 
după datele de călătorie. Un bilet de călătorie în direcție / în continuare este necesar pentru a dovedi durata 
șederii dumneavoastră.
Input context (CCI > 0.001):     Ecuador requires that Cuban citizens receive an invitation letter prior to 
entering Ecuador through international airports or frontier admission points. This letter must be legalized by 
the(0.002) Ecuadorian Ministry of Foreign Affairs, and comply with certain requirements. These requirements are 
designed to provide an organized migratory flow between both countries. Cuban citizens who are US green card 
holders should visit an(0.001) Ecuadorian Consulate to obtain an exemption to this requirement(0.003).(0.002)
Output context (CCI > 0.001):    Ecuadorul cere ca cetățenii Cubei să primească o scrisoare de invitație înainte de
a intra în Ecuador folosind aeroporturile internaționale sau punctele de trecere a frontierei. Scrisoarea aceasta 
trebuie să fie legalizată de către Ministerul Afacerilor Externe al Ecuadorului și trebuie să respecte anumite 
cerințe. Cerințele acestea sunt concepute pentru a asigura o deplasare bine organizată între cele două țări. 
Cetățenii cubanezi care dețin cărți verzi pentru SUA trebuie să viziteze un consulat ecuadorian pentru a obține o 
scutire de la această cerință.(0.006)
#2.
Generated output (CTI > 0.165):  Pașaportul dumneavoastră trebuie să fie valabil pentru cel puțin 6 luni după 
datele de călătorie. Un(1.225) bilet de călătorie în direcție / în continuare este necesar pentru a dovedi durata 
șederii dumneavoastră.
Input context (CCI > 0.001):     Ecuador requires that Cuban citizens receive an invitation letter prior to 
entering Ecuador through international airports or frontier admission points. This letter must be legalized by 
the(0.001) Ecuadorian Ministry of Foreign Affairs, and comply with certain requirements. These requirements are 
designed to provide an organized migratory flow between both countries. Cuban citizens who are US green card 
holders should visit an(0.001) Ecuadorian Consulate to obtain an exemption to this requirement(0.001).
Output context (CCI > 0.001):    Ecuadorul cere ca cetățenii Cubei să primească o scrisoare de invitație înainte de
a int

Context with contextual cues (std λ=2.00, top 5) followed by output sentence with context-sensitive target spans 
(all)
(CTI = "kl_divergence", CCI = "saliency" w/ "probability" target)

Input context:  There are many things you have to take into consideration before and when you travel somewhere.
Input current: When you travel, expect things not to be like they are "back home". Manners, laws, food, traffic, 
lodging, standards, language and so on will to some extent differ from where you live.
Output context: Sunt multe lucruri de luat în considerare înainte și în timpul călătoriei spre o anumită 
destinație.
Output current: Când călătorești, așteaptă să nu fie ca "în casă".

#1.
Generated output (CTI > 0.124):  Când(0.901) călătorești, așteaptă să nu fie ca "în casă".
Input context (CCI > 0.010):     There are many things you have to take into consideration before and when you 
travel somewhere(0.011).
Output context (CCI > 0.011):    Sunt multe lucruri de luat în considerare înainte și în timpul călătoriei spre o 
anumită destinație.(0.024)
#2.
Generated output (CTI > 0.124):  Când călătorești(0.667), așteaptă să nu fie ca "în casă".
Input context (CCI > 0.007):     There are many things you have to take into consideration before and when you 
travel somewhere.
Output context (CCI > 0.008):    Sunt multe lucruri de luat în considerare înainte și(0.015) în timpul călătoriei 
spre o anumită destinație.

Context with contextual cues (std λ=2.00, top 5) followed by output sentence with context-sensitive target spans 
(all)
(CTI = "kl_divergence", CCI = "saliency" w/ "probability" target)

Input context:  At home, due to this constant exposure to the local germs, odds are very high that you're already 
immune to them.
Input current: But in other parts of the world, where the bacteriological fauna are new to you, you're far more 
likely to run into problems.
Output context: Acasă, datorită expunerii constante la microbii locali, șansele de a fi deja imun la ei sunt foarte
mari.
Output current: Dar în alte părți ale lumii, unde fauna bacteriologică este nouă pentru tine, e mult mai probabil 
să ai probleme.

#1.
Generated output (CTI > 0.045):  Dar în alte părți(0.534) ale lumii, unde fauna bacteriologică este nouă pentru 
tine, e mult mai probabil să ai probleme.
Input context (CCI > 0.003):     At home, due to this constant exposure to the local germs, odds are very high that
you're already immune to them.
Output context (CCI > 0.003):    Acasă, datorită expunerii constante la microbii locali, ș(0.009)ansele de a fi 
deja imun la ei sunt foarte mari.
#2.
Generated output (CTI > 0.045):  Dar(0.389) în alte părți ale lumii, unde fauna bacteriologică este nouă pentru 
tine, e mult mai probabil să ai probleme.
Input context (CCI > 0.004):     At home, due to this constant exposure to the local germs, odds are very high that
you're already immune to them(0.005).
Output context (CCI > 0.004):    Acasă, datorită expunerii constante la microbii locali, șansele de a fi deja imun 
la ei sunt foarte mari.(0.009)
#3.
Generated output (CTI > 0.045):  Dar în alte părți ale lumii, unde fauna bacteriologică este nouă pentru tine, 
e(0.057) mult mai probabil să ai probleme.
Input context (CCI > 0.001):     At home, due to this constant exposure to the local germs, odds are very high that
you're already immune to them.
Output context (CCI > 0.001):    Acasă, datorită expunerii constante la microbii locali, ș(0.002)ansele de a fi 
deja imun la ei sunt foarte mari.

Context with contextual cues (std λ=2.00, top 5) followed by output sentence with context-sensitive target spans 
(all)
(CTI = "kl_divergence", CCI = "saliency" w/ "probability" target)

Input context:  Fill your home with a rich coffee in the morning and some relaxing chamomile tea at night. When 
you're on a staycation, you have the time to treat yourself and take a few extra minutes to brew up something 
special. If you're feeling more adventurous, take the opportunity to juice or blend up some smoothies:
Input current: maybe you'll discover a simple beverage that you can make for breakfast when you're back to your 
daily routine.
Output context: Umple-ți casa cu o aromă de cafea savuroasă dimineața și de ceai de mușețel relaxant seara. Când 
stați consemnat la domiciliu, aveți timp să vă faceți plăcerile și să vă permiteți câteva minute extra ca să 
pregătiți o masă specială. Dacă ai chef de aventură, profită de ocazie ca să faci un suc sau să amesteci niște 
smoothies:
Output current: poate vei descoperi o băutură simplă pe care o poți face pentru micul dejun când te-ai întors la 
rutina de zi cu zi.

#1.
Generated output (CTI > 0.161):  poate(2.328) vei descoperi o băutură simplă pe care o poți face pentru micul dejun
când te-ai întors la rutina de zi cu zi.
Input context (CCI > 0.001):     Fill your home with a rich coffee in the morning and some relaxing chamomile tea 
at night. When you're on a staycation, you have the time to treat yourself and take a few extra minutes to brew up 
something special. If you're feeling more adventurous, take the opportunity to juice or blend up some 
smoothies(0.003):(0.003)
Output context (CCI > 0.001):    Umple-ți casa cu o aromă de cafea savuroasă dimineața și de ceai de mușețel 
relaxant seara. Când stați consemnat la domiciliu, aveți timp să vă faceți plăcerile și să vă permiteți câteva 
minute extra ca să pregătiți o masă specială. Dacă ai chef de aventură, profită de ocazie ca să faci un suc sau să 
amesteci niște smoothies:(0.002)
#2.
Generated output (CTI > 0.161):  poate vei descoperi o băutură simplă pe care o po(1.392)ți face pentru micul dejun
când te-ai întors la rutina de zi cu zi.
Input context (CCI > 0.002):     Fill your home with a rich coffee in the morning and some relaxing chamomile tea 
at night. When you're on a staycation, you have the time to treat yourself and take a few extra minutes to brew up 
something special. If you're feeling more adventurous, take the opportunity to juice or blend up some smoothies:
Output context (CCI > 0.002):    Umple-ți casa cu o aromă de cafea savuroasă dimineața și de ceai de mușețel 
relaxant seara. Când stați(0.003) consemnat la domiciliu, ave(0.003)ți(0.005) timp să vă faceți plăcerile și să vă 
permiteți câteva minute extra ca să pregătiți o masă specială. Dacă ai chef de aventură, profită de ocazie ca să 
faci un suc sau să am(0.004)este(0.005)ci niște smoothies:
#3.
Generated output (CTI > 0.161):  poate vei descoperi o băutură simplă pe care o poți face pentru micul dejun când 
te-(0.355)ai întors la rutina de zi cu zi.
Input context (CCI > 0.002):     Fill your home with a rich coffee in the morning and some relaxing chamomile tea 
at night. When you're on a staycation, you have the time to treat yourself and take a few extra minutes to brew up 
something special. If you're feeling more adventurous, take the opportunity to juice or blend up some smoothies:
Output context (CCI > 0.002):    Umple-ți casa cu o aromă de(0.004) cafea savuroasă dimineața și(0.002) de ceai de 
mușețel relaxant seara. Când stați consemnat la domiciliu, aveți timp să vă faceți plăcerile și să vă permiteți 
câteva minute extra ca să pregătiți o masă specială. Dacă ai chef de(0.005) aventură, profită de ocazie ca să faci 
un suc sau să amesteci ni(0.003)ște smoothies:
#4.
Generated output (CTI > 0.161):  poate vei(0.332) descoperi o băutură simplă pe care o poți face pentru micul dejun
când te-ai întors la rutina de zi cu zi.
Input context (CCI > 0

Context with contextual cues (std λ=2.00, top 5) followed by output sentence with context-sensitive target spans 
(all)
(CTI = "kl_divergence", CCI = "saliency" w/ "probability" target)

Input context:  In summer, especially, you'll need to watch out for mosquitoes if you decide to hike through the 
rainforest.
Input current: Even if you're driving through the subtropical rainforest, a few seconds with the doors open while 
you get inside the vehicle is enough time for mosquitoes to get in the vehicle with you.
Output context: Vara, în special, va trebui să ai grijă la țânțari, dacă decizi să faci un traseu prin pădurea 
tropicală.
Output current: Chiar dacă conduci prin pădurea tropicală subtropicală, câteva secunde cu ușile deschise în timp ce
intri în vehicul este suficient timp pentru țânțari să intre în vehicul cu tine.

#1.
Generated output (CTI > 0.121):  Chiar dacă conduci prin pădurea trop(1.747)icală subtropicală, câteva secunde cu 
ușile deschise în timp ce intri în vehicul este suficient timp pentru țânțari să intre în vehicul cu tine.
Input context (CCI > 0.002):     In summer, especially, you'll need to watch out for mosquitoes if you decide to 
hike through the rainforest.
Output context (CCI > 0.002):    Vara, în special, va trebui să ai grijă la țânțari, dacă decizi să faci un traseu 
prin pădurea(0.002) trop(0.004)icală.
#2.
Generated output (CTI > 0.121):  Chiar dacă conduci prin pădurea tropicală subtropicală, câteva secunde cu ușile 
deschise în timp ce intri în vehicul este suficient timp pentru ț(1.416)ânțari să intre în vehicul cu tine.
Input context (CCI > 0.005):     In summer, especially, you'll need to watch out for mosquitoes if you decide to 
hike through the rainforest.
Output context (CCI > 0.005):    Vara, în special, va trebui să ai grijă la ț(0.016)ânțari, dacă decizi să faci un 
traseu prin pădurea tropicală.
#3.
Generated output (CTI > 0.121):  Chiar dacă conduci prin pădurea tropicală sub(0.470)tropicală, câteva secunde cu 
ușile deschise în timp ce intri în vehicul este suficient timp pentru țânțari să intre în vehicul cu tine.
Input context (CCI > 0.003):     In summer, especially, you'll need to watch out for mosquitoes if you decide to 
hike through the rainforest.
Output context (CCI > 0.004):    Vara, în special, va trebui să ai grijă la țânțari, dacă decizi să faci un traseu 
prin pădurea(0.004) trop(0.006)ical(0.004)ă.
#4.
Generated output (CTI > 0.121):  Chiar dacă conduci prin pădurea tropicală subtropicală, câteva secunde cu ușile 
deschise în timp ce intri în vehicul este suficient timp pentru țânțari(0.455) să intre în vehicul cu tine.
Input context (CCI > 0.003):     In summer, especially, you'll need to watch out for mosquitoes if you decide to 
hike through the rainforest.
Output context (CCI > 0.003):    Vara, în special, va trebui să ai grijă la ț(0.006)ân(0.004)țari(0.006), dacă 
decizi să faci un traseu prin pădurea tropicală.
#5.
Generated output (CTI > 0.121):  Chiar(0.307) dacă conduci prin pădurea tropicală subtropicală, câteva secunde cu 
ușile deschise în timp ce intri în vehicul este suficient timp pentru țânțari să intre în vehicul cu tine.
Input context (CCI > 0.008):     In summer, especially, you'll need to watch out for mosquitoes if you decide to 
hike through the rainforest.
Output context (CCI > 0.008):    Vara, în special, va trebui să ai grijă la țânțari, dacă decizi să faci un traseu 
prin pădurea(0.009) tropicală.(0.017)
#6.
Generated output (CTI > 0.121):  Chiar dacă condu(0.268)ci prin pădurea tropicală subtropicală, câteva secunde cu 
ușile deschise în timp ce intri în vehicul este suficient timp pentru țânțari să intre în vehicul cu tine.
Input context (CCI > 0.002):     In summer, especially, you'll need to watch out for mosquitoes if you decide to 
hike through the rainforest.
Output context (CCI > 0.002):    Vara, în special, va trebui să ai grijă la țânțari, dacă(0.002) deci(0.003)zi să 
faci un traseu prin pădurea tropicală.
#7.
Generated out

Context with contextual cues (std λ=2.00, top 5) followed by output sentence with context-sensitive target spans 
(all)
(CTI = "kl_divergence", CCI = "saliency" w/ "probability" target)

Input context:  Courier companies are well paid for delivering things quickly. Frequently, time is very important 
with business documents, merchandise or spare parts for an urgent repair. On some routes, the larger companies have
their own planes, but for other routes and smaller firms there was a problem. If they sent things by air freight, 
on some routes it may have taken days to get through unloading and customs.
Input current: The only way to get it through faster was to send it as checked luggage. Airline regulations will 
not allow them to send luggage without a passenger, which is where you come in.
Output context: Companiile de curierat sunt bine plătite pentru livrările rapide. Adesea, timpul de livrare este 
esențial în ceea ce privește documentele, mărfurile sau piesele de schimb pentru o reparație urgentă. Pe unele 
rute, companiile mai mari dețin propriile avioane, dar alte rute și companiile mai mici aveau o problemă. Dacă au 
trimis ceva pe calea aerului, pe unele rute poate dura zile întregi descărcarea și vămuirea.
Output current: Singura modalitate de a trece mai repede a fost să o trimiteți ca bagaj înregistrat. Reglementările
companiei aeriene nu le permit să trimită bagaje fără pasager, care este locul în care intri.

#1.
Generated output (CTI > 0.174):  Singura modalitate de a trece mai repede a fost să o trimiteți(2.056) ca bagaj 
înregistrat. Reglementările companiei aeriene nu le permit să trimită bagaje fără pasager, care este locul în care 
intri.
Input context (CCI > 0.001):     Courier companies are well paid for delivering things quickly. Frequently, time is
very important with business documents, merchandise or spare parts for an urgent repair. On some routes, the larger
companies have their own planes, but for other routes and smaller firms there was a problem. If they sent things by
air freight, on some routes it may have taken days to get through unloading and customs.
Output context (CCI > 0.001):    Companiile de curierat sunt bine plătite pentru livrările rapide. Adesea, timpul 
de livrare este esențial în ceea ce privește(0.001) documentele, mărfurile sau piesele de schimb pentru o reparație
urgentă. Pe unele rute, companiile mai mari dețin(0.001) propriile avioane, dar alte rute și companiile mai mici 
aveau o problemă. Dacă au trimis ceva pe calea aerului, pe unele rute poate dura zile între(0.001)gi descărcarea 
și(0.001) vă(0.003)muirea.
#2.
Generated output (CTI > 0.174):  Sing(1.870)ura modalitate de a trece mai repede a fost să o trimiteți ca bagaj 
înregistrat. Reglementările companiei aeriene nu le permit să trimită bagaje fără pasager, care este locul în care 
intri.
Input context (CCI > 0.003):     Courier companies are well paid for delivering things quickly. Frequently, time is
very important with business documents, merchandise or spare parts for an urgent repair. On some routes, the larger
companies have their own planes, but for other routes and smaller firms there was a problem. If they sent things by
air freight, on some routes it may have taken days to get through unloading and customs(0.004).
Output context (CCI > 0.004):    Companiile de curierat sunt bine plătite pentru livrările rapide. Adesea, timpul 
de livrare este esențial în ceea ce privește documentele, mărfurile sau piesele de schimb pentru o reparație 
urgentă. Pe unele rute, companiile mai mari dețin propriile avioane, dar alte rute și companiile mai mici aveau o 
problemă. Dacă au trimis ceva pe calea aerului, pe unele rute poate dura zile întregi descărcarea și 
vă(0.005)muirea(0.006).(0.016)
#3.
Generated output (CTI > 0.174):  Singura modalitate de a trece mai repede a fost să o trimiteți ca bagaj 
înregistrat.(1.067) Reglementările companiei aeriene nu le permit să trimită bagaje fără pasager, care este locul 
în care

# **Analysis of the 6 common sentences across all sentences**
**Sentence #12**


**Generated outputs**: 12\
**Analysis**:  We can observe that the CTI highlighted the pronoun "dumneavoastră" in two instances, one for each sentence of the generated output. In each situation, the almost all highlighted contextual cues (determiners, nouns, word-parts) are not relevant to suggesting a formal pronoun, as generated. This outcome could be explained by the lack of usage of any second person pronouns in the context. However, certain words could imply the usage of a formal versus an informal form, but PECoRe usually fails to detect these tokens, except for the "Afa(0.001)cerilor"(Affairs). Although this token suggests formality, it has a very low score attributed to it, which is in line with the other tokens identified by the CCI step. This fact indicates the model is "coming to the right conclusion for the wrong reasons".

**Sentence #18**

**Generated outputs**: 2\
**Analysis**: In this example, we can observe a lack of contextual cues, the only ones being represented by the expression "în casă" (wrongly translated) and the conjunction și (and). None of these tokens are contributing to the generation of the context-sensitive target span represented by the verb "călătorești(0.667)" (travel) which implies the informal singular form of you by the suffix "-ști:. The informal form is also maintained for the rest of the sentence, suggested by the verb "așteaptă" (wait). Therefore, similar to the previous case, the model provides a correct translation, but does not rely on context for generating it.

**Sentence #25**

**Generated outputs**: 3\
Analysis: Similar to sentence 18, we can observe a shortage of contextual cues in this case. Moreover, we can observe that the only token highlighted by CTI is the verb "e(0.057)", which is the contracted form of the verb "este" (is), suggesting an informal setting. This is supported by the translation of  "you" to "tine" and the verb form "să ai" (to have). Nevertheless, the model does not manage to justify its decision even though the context contains the verb "a fi" (to be), which is specific to the informal singular form of you.


**Sentence #29**

**Generated outputs**: 4\
**Analysis**: This instance is the first one in which we can observe that the model rightfully justifies its generated output concerning the phenomenon we are interested in. Two verbs are highlighted by the CTI, "po(1.392)ți" (can) and "vei(0.332)" (will) that indirectly indicate the informal singular form of "you". In each case, the contextual cues are represented by the verbs that suggest the same pronoun form. Although the generated output contains further evidence for the informal singular form of "you" ("te-ai întors"), this is not detected by CTI. This might be due to the fact that it already captured the former instances and this later instance is dependent on them.

**Sentence #39**

**Generated outputs**: 7\
**Analysis**: Akin to the previous sentence, the informal singular form of you is indicated indirectly through the verb "condu(0.268)ci" (drive), which relies on the contextual cue "deci(0.003)zi" (decide), which indirectly suggests the informal singular form of "you". Other verb forms could have contributed to the generated output but are not captured by either CTI or CCI.

**Sentence #53**

**Generated outputs**: 7
**Analysis**: In this case, CTI highlighted the verb "trimiteți(2.056)" (send) that suggests a formal version of the pronoun "you". However, all the highlighted contextual cues are not relevant to suggesting a formal pronoun, as generated. Among these contextual clues is the part-word "vă(0.003)muirea" (customs) which, taken separately as "vă" (you), could suggest the generated output. Regardless if the model interprets vă as you as indicative of the formal tone, it relies on incorrect context to derive its conclusion.
